docs/concurrency.org

```diff
@@ -1,68 +1,251 @@
+  WaitForSingleObject(hThread, INFINITE);
+
+  // Close the thread handle
+  CloseHandle(hThread);
+#else
+  pthread_t thread;
+  int rc;
+
+  rc = pthread_create(&thread, NULL, ThreadFunc, NULL);
+  if (rc) {
+    printf("Failed to create thread. Return code: %d\n", rc);
+    return 1;
+  }
+
+  // Wait for the thread to finish
+  pthread_join(thread, NULL);
+#endif
 
-*** Old async/await
-Using old async/await from before 0.11 (not recommanded because unstable !!! https://github.com/ziglang/zig/issues/6025 and might probably never get back in the language https://ziglang.org/download/0.12.0/release-notes.html#AsyncAwait-Feature-Status)
+  printf("Everything is done.\n");
+
+  return 0;
+}
+#+end_src
+
+Or you could write your own wrapper kind of like the way Zig does (this is not going to work on Windows, but you get the idea):
+#+begin_src c
+#include <stdio.h>
+
+#ifdef _WIN32
+#include <windows.h>
+#else
+#include <pthread.h>
+#endif
+
+int myCreate(unsigned long *thread, void *func) {
+#ifdef _WIN32
+  return hThread = CreateThread(NULL, 0, func, NULL, 0, thread);
+#else
+  return pthread_create(thread, NULL, func, NULL);
+#endif
+}
+
+void myJoin(unsigned long thread) {
+#ifdef _WIN32
+  return WaitForSingleObject(thread, INFINITE);
+#else
+  pthread_join(thread, NULL);
+#endif
+}
+
+void *ThreadFunc(void *arg) {
+  printf("Thread running...\n");
+  return NULL;
+}
+
+int main() {
+  pthread_t thread; // TODO I should also wrap that which is pthread specific
+
+  int rc = myCreate(&thread, ThreadFunc);
+  if (rc) {
+    printf("Failed to create thread. Return code: %d\n", rc);
+    return 1;
+  }
+
+  myJoin(thread);
+
+  printf("Everything is done.\n");
+
+  return 0;
+}
+#+end_src
+
+**** Thread synchronization
+
+
+**** Leaky abstraction
+TODO: est ce que je peux fine tuned threads en Zig ? (priority, stack size, ...)
+
+**** Conclusion
+Zig threads are really useful since they have a very user-friendly abstraction with not a lot of functionalites that are almost never used anyway. This abstraction is also very useful for what we saw earlier, you don't have to worry about the target system, Zig is going to choose the right implementation for you.
+
+But this leaky abstraction comes at a cost, you can not fine-tune your threads as much as you would like to.
+
+If you need specific thread functionalities, like the ones we talked about, you can still do that in Zig by wrapping the C pthread library for exemple or directly use the OS native threads you want.
+
+*** async/await
+This method uses suspensible stackless coroutines, this solution does not necessarly mean that you are going to have multiple threads or parallelism.
+
+We are not going to dive deeper into this solution because it has been deprecated since 0.11 and is not coming back soon.
+
+However it is still a good reading and way to understand concurency to read this [[https://zig.guide/async/introduction/][very good guide]] that was made for this solution.
```

Comment:
Add citation in citeproc

Reply:
https://github.com/Pismice/HEIG_ZIG/pull/11#discussion_r1607147705